daemons/based/based_transaction.c

```diff
@@ -14,270 +14,59 @@
-{
-    struct request_data *data = NULL;
-    GQueue *transaction = NULL;
-    int rc = pcmk_rc_ok;
-
-    CRM_ASSERT((client != NULL) && (request != NULL));
-
-    transaction = get_transaction(client);
-    if (transaction == NULL) {
-        return pcmk_rc_no_transaction;
-    }
-
-    rc = validate_transaction_request(request);
-    if (rc != pcmk_rc_ok) {
-        return rc;
-    }
-
-    data = create_request_data(client, request, privileged);
-    if (data == NULL) {
-        return ENOMEM;
-    }
-
-    g_queue_push_tail(transaction, data);
-    return pcmk_rc_ok;
-}
-
-/*!
- * \internal
- * \brief Free a CIB client's transaction (if any) and its requests
- *
- * \param[in] client  Client whose transaction to discard
- */
-void
-based_discard_transaction(const pcmk__client_t *client)
-{
-    bool found = false;
-
-    CRM_ASSERT(client != NULL);
-
-    if (transactions != NULL) {
-        found = g_hash_table_remove(transactions, client->id);
-    }
-
-    crm_trace("%s for client %s (%s)",
-              (found? "Found and removed transaction" : "No transaction found"),
-              pcmk__client_name(client),
-              pcmk__s(client->id, "unidentified"));
-}
-
 /*!
  * \internal
  * \brief Process requests in a transaction
  *
  * Stop when a request fails or when all requests have been processed.
  *
  * \param[in,out] transaction  Transaction to process
- * \param[in]     client_name  Client name (for logging only)
- * \param[in]     client_id    Client ID (for logging only)
+ * \param[in]     client       CIB client
  *
  * \return Standard Pacemaker return code
  */
 static int
-process_transaction_requests(GQueue *transaction, const char *client_name,
-                             const char *client_id)
+process_transaction_requests(xmlNodePtr transaction, const pcmk__client_t *client)
 {
-    for (struct request_data *data = g_queue_pop_head(transaction);
-         data != NULL; data = g_queue_pop_head(transaction)) {
-
-        const char *op = crm_element_value(data->request, F_CIB_OPERATION);
-
-        int rc = cib_process_request(data->request, data->privileged,
-                                     data->client);
+    const char *client_name = pcmk__client_name(client);
+    const char *client_id = pcmk__s(client->id, "unidentified");
+
+    for (xmlNodePtr request = first_named_child(transaction, T_CIB_COMMAND);
+         request != NULL; request = crm_next_same_xml(request)) {
+
+        const char *op = crm_element_value(request, F_CIB_OPERATION);
+        const char *host = crm_element_value(request, F_CIB_HOST);
+        const cib__operation_t *operation = NULL;
+        int rc = pcmk_rc_ok;
+
+        cib__get_operation(op, &operation);
+
+        // Client should have rejected an unsupported request
+        CRM_CHECK((operation != NULL)
+                  && pcmk_is_set(operation->flags, cib__op_attr_transaction)
+                  && (host == NULL),
+                  rc = EOPNOTSUPP);
```

Comment:
The CIB is proxied to Pacemaker Remote nodes, so we can't assume the client and server came from the same Pacemaker version

Reply:
How are we assuming that here?

Transactions haven't made it into a release yet, luckily. I'm only assuming the client isn't running a version that uses the experimental server-based implementation of transactions. I think that's reasonable since it was only in the main branch and we quickly introduced a controller regression based on it.

Reply:
Future-proofing. We can't assume (for all future combinations of client and server versions) that the client has rejected an unsupported request. The client may be running an older or newer version with a different conception of supported than the server.

Not a big deal since we're re-validating here, but probably doesn't merit a CHECK